File: interop-tests/test_shake.cpp
/**
 * Test SHAKE256 output to compare with JS
 */

#include <rescue/utils.hpp>
#include <rescue/field.hpp>

#include <iomanip>
#include <iostream>
#include <sstream>

using namespace rescue;

std::string to_hex(const uint8_t* data, size_t len) {
    std::stringstream ss;
    for (size_t i = 0; i < len; i++) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(data[i]);
    }
    return ss.str();
}

std::string fp_to_hex(const Fp& val) {
    auto arr = val.to_bytes();
    std::stringstream ss;
    for (auto b : arr) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(b);
    }
    return ss.str();
}

int main() {
    std::cout << "=== SHAKE256 Test (C++) ===\n\n";

    // Test with the same initialization as cipher mode
    Shake256 hasher;
    hasher.update("encrypt everything, compute anything");

    // We need to squeeze all at once - our implementation doesn't support incremental XOF
    // Extract 3 * 48 = 144 bytes
    auto output = hasher.finalize(144);

    std::cout << "First 3 chunks (48 bytes each):\n\n";
    std::cout << "chunk #1: " << to_hex(output.data(), 48) << "\n";
    std::cout << "chunk #2: " << to_hex(output.data() + 48, 48) << "\n";
    std::cout << "chunk #3: " << to_hex(output.data() + 96, 48) << "\n";

    // Now get field elements using wide_bytes_to_fp logic
    std::cout << "\nFirst 5 field elements from XOF (using wide reduction):\n\n";
    
    Shake256 hasher2;
    hasher2.update("encrypt everything, compute anything");
    auto output2 = hasher2.finalize(5 * 48);

    for (int i = 0; i < 5; i++) {
        std::span<const uint8_t> chunk(output2.data() + i * 48, 48);
        
        // Implement wide_bytes_to_fp logic inline
        std::array<uint64_t, 8> wide_limbs = {0};
        for (size_t j = 0; j < 48; ++j) {
            wide_limbs[j / 8] |= static_cast<uint64_t>(chunk[j]) << ((j % 8) * 8);
        }
        
        uint256 low{wide_limbs[0], wide_limbs[1], wide_limbs[2], wide_limbs[3]};
        uint256 high{wide_limbs[4], wide_limbs[5], 0, 0};
        
        Fp elem = Fp(low) + Fp(high) * Fp(uint64_t{38});
        std::cout << "  elem[" << i << "] = " << fp_to_hex(elem) << "\n";
    }

    std::cout << "\n=== Done ===\n";
    return 0;
}

File: interop-tests/test_shake.js
/**
 * Test SHAKE256 output to compare with C++
 */

const { shake256 } = require('@noble/hashes/sha3.js');

function toHex(arr) {
    return Array.from(arr).map(b => b.toString(16).padStart(2, '0')).join('');
}

console.log('=== SHAKE256 Test (JavaScript) ===\n');

// Test with the same initialization as cipher mode
const hasher = shake256.create({ dkLen: 256 / 8 });
hasher.update(new TextEncoder().encode('encrypt everything, compute anything'));

// Extract 48 bytes at a time (like JS library does)
console.log('First 3 XOF outputs (48 bytes each):\n');

const out1 = hasher.xof(48);
console.log('xof(48) #1:', toHex(out1));

const out2 = hasher.xof(48);
console.log('xof(48) #2:', toHex(out2));

const out3 = hasher.xof(48);
console.log('xof(48) #3:', toHex(out3));

// Also show what the first few field elements would be
const P = 2n ** 255n - 19n;

function deserializeLE(bytes) {
    let value = 0n;
    for (let i = 0; i < bytes.length; i++) {
        value |= BigInt(bytes[i]) << BigInt(i * 8);
    }
    return value;
}

function bigintToHex(val) {
    const bytes = [];
    let temp = val;
    for (let i = 0; i < 32; i++) {
        bytes.push(Number(temp & 0xffn));
        temp >>= 8n;
    }
    return bytes.map(b => b.toString(16).padStart(2, '0')).join('');
}

// Start fresh to get field elements
const hasher2 = shake256.create({ dkLen: 256 / 8 });
hasher2.update(new TextEncoder().encode('encrypt everything, compute anything'));

console.log('\nFirst 5 field elements from XOF:\n');
for (let i = 0; i < 5; i++) {
    const bytes = hasher2.xof(48);
    const value = deserializeLE(bytes) % P;
    console.log(`  elem[${i}] = ${bigintToHex(value)}`);
}

console.log('\n=== Done ===');
